### PR TITLE
Network information endpoint - implementation

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#819 

# Overview

- [x] I have implemented the `/network/information` endpoint
- [x] I have provided an integration test scenario for it.

# Comments

### CODE REVIEW

**Review by rvl — APPROVED** [review]
It looks like it meets the spec in the schema, and is tested too. :+1:

But I have to say the swagger spec is a bit weird. Maybe we can fix them later, if Daedalus are depending on the current schema.
- There needn't be a distinction between "ready" and "restoring". Just use a percentage and make sure to round down. Then 100% means ready.
- Why the quantity+unit thing? Is the client supposed to parse the unit and scale the quantity accordingly? I doubt it. Use `{ "sync_progress_percent": 99 }`. Even better, express sync progress as wallet block height and network block height.
- If NTP and update system information are unavailable, simply omit them. Javascript apps deal with null and undefined all the time.

**Comment by KtorZ** [comment]
@rvl Thanks for the review and feedback! To answer a few of your questions:

> **There needn't be a distinction between "ready" and "restoring". Just use a percentage and make sure to round down. Then 100% means ready.**

Actually, there's no particular distinction for us and I'd call that "an implementation detail". This is because we are so-to-speak, always restoring and chasing the network tip. Yet, Daedalus and clients in general like to see a clear change in state, otherwise, they feel the software is stuck and that something is wrong. To be honest, I don't mind either way although the "ready" state change is self documenting.

> **Why the quantity+unit thing? Is the client supposed to parse the unit and scale the quantity accordingly? I doubt it. Use { "sync_progress_percent": 99 }.**

Haha :) ... You probably haven't experienced the pain of repeating 12 times what's the unit for amount and what's the unit for that timestamp, even though it's written all over the documentation. Having the unit as part of the response makes the API self documenting and makes for clean structure. `sync_progress_percent` would work, but I would argue that it's definitely less elegant than: `{ "sync_progress": { "quantity": 99, "unit": "percent" }}`. We aren't expecting anything about clients, but it's a common format we use all over the API, so it makes it very consistent for quantified numbers.

> **Even better, express sync progress as wallet block height and network block height.**

That was actually my first design before :upside_down_face: ... Leave the computation to the client but Daedalus insisted to have a percentage value. So here we go. Note however that I am going to add the block height in another PR (as per the decision plan in the ticket), so basically both pieces of information will be available.

> **If NTP and update system information are unavailable, simply omit them.**

Yeah :/ .. I wanted to prepare ground for documentation a bit here but, it turns out to be more confusing than helpful I think. I'll remove that.

File: lib/core-integration/src/Test/Integration/Scenario/API/Network.hs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    , protocolUpdates
+    , request
+    , syncProgress
+    , verify
+    )
+
+spec :: forall t. SpecWith (Context t)
+spec = do
+    it "NETWORK - Can query network information" $ \ctx -> do
+        let endpoint = ("GET", "v2/network/information")
```
Comment (piotr-iohk) [inline lib/core-integration/src/Test/Integration/Scenario/API/Network.hs:31]: A tests against `405` could be also added, like https://github.com/input-output-hk/cardano-wallet/blob/f0f721bdaff148cab01299334833a583b01a8119/lib/core-integration/src/Test/Integration/Scenario/API/Wallets.hs#L855

**Comment by rvl** [comment]
I did not see that there is a tolerance of 5 blocks in the Ready/Restoring calculation. So actually, it does make sense to distinguish between the two.

**Comment by KtorZ** [comment]
> I did not see that there is a tolerance of 5 blocks in the Ready/Restoring calculation. So actually, it does make sense to distinguish between the two.

Yes that too. Otherwise the wallet is back in "restore 99%" basically every slot. The `5` is pretty much arbitrary though, but we should have something that plays well with Praos.

File: lib/core-integration/src/Test/Integration/Scenario/API/Network.hs (hunk 1 above)
Comment (KtorZ) [reply lib/core-integration/src/Test/Integration/Scenario/API/Network.hs:31]: will do in a separate PR (with the CLI)

**Pull request merged**